boa/src/builtins/date/tests.rs

```diff
@@ -0,0 +1,1428 @@
+    )
+    .expect("Successful eval");
+    assert_eq!(Value::string("Wed, 08 Jul 2020 09:16:15 GMT"), actual);
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_to_iso_string() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "let dt = new Date(Date.UTC(2020, 06, 08, 09, 16, 15, 779)); dt.toISOString()",
+    )
+    .expect("Successful eval");
+    assert_eq!(Value::string("2020-07-08T09:16:15.779Z"), actual);
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_to_json() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "let dt = new Date(Date.UTC(2020, 06, 08, 09, 16, 15, 779)); dt.toJSON()",
+    )
+    .expect("Successful eval");
+    assert_eq!(Value::string("2020-07-08T09:16:15.779Z"), actual);
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_to_string() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "let dt = new Date(2020, 06, 08, 09, 16, 15, 779); dt.toString()",
+    )
+    .ok();
+
+    assert_eq!(
+        Some(Value::string(
+            Local::now()
+                .format("Wed Jul 08 2020 09:16:15 GMT%:z")
+                .to_string()
+        )),
+        actual
+    );
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_to_time_string() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "let dt = new Date(2020, 06, 08, 09, 16, 15, 779); dt.toTimeString()",
+    )
+    .ok();
+
+    assert_eq!(
+        Some(Value::string(
+            Local::now().format("09:16:15 GMT%:z").to_string()
+        )),
+        actual
+    );
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_to_utc_string() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "let dt = new Date(Date.UTC(2020, 06, 08, 09, 16, 15, 779)); dt.toUTCString()",
+    )
+    .expect("Successful eval");
+    assert_eq!(Value::string("Wed, 08 Jul 2020 09:16:15 GMT"), actual);
+
+    Ok(())
+}
+
+#[test]
+fn date_proto_value_of() -> Result<(), Box<dyn std::error::Error>> {
+    let realm = Realm::create();
+    let mut engine = Interpreter::new(realm);
+
+    let actual = forward_val(
+        &mut engine,
+        "new Date(Date.UTC(2020, 06, 08, 09, 16, 15, 779)).valueOf()",
+    )
+    .expect("Successful eval");
+    assert_eq!(Value::number(1594199775779f64), actual);
+
+    Ok(())
+}
+
+#[test]
+fn date_neg() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
It looks like `Value::Date` is never used, because `Date` is a constructed type (unlike the primitives). This means that all the JSON and primitive logic goes unused, and e.g. `Object([Slot] = Date)` always becomes `NaN`.

Should I just delete `Value::Date` and update the `Value::Object` callsites to also inspect the internal slot?

Affects this test and the following test.

Reply:
One option would be to implement `TryFrom<ObjectData> for Value`, and recursively handle `Object` that way?

Reply:
> It looks like `Value::Date` is never used, because `Date` is a constructed type (unlike the primitives)
> Should I just delete `Value::Date`.

Yes. we should delete the `Value::Date`as it is not a primitive. 

Im not sure what you mean by:
> update `Value::Object` callsites to also inspect the internal slot?

The Date object internal slots should be stored as native types in `ObjectData`

Reply:
> One option would be to implement `TryFrom<ObjectData> for Value`, and recursively handle `Object` that way?

Could you explain a bit more about this, why do we need and what we are trying to solve